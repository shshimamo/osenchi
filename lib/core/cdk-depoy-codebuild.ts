import * as cdk from 'aws-cdk-lib';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as iam from 'aws-cdk-lib/aws-iam';

export class CdkDeployCodeBuildStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const buildImage = codebuild.LinuxBuildImage.fromDockerRegistry(
            "public.ecr.aws/bitnami/node:16.18.1"
        )
        const project = new codebuild.Project(this, "OsenchiProject", {
            projectName: "osenchi-cdk-deploy",
            environment: {
                buildImage
            },
            buildSpec: codebuild.BuildSpec.fromObject({
                version: "0.2",
                phases: {
                    install: {
                        commands: [
                            'npm install -g aws-cdk',
                            "npm install",
                        ]
                    },
                    pre_build: {
                       commands: [
                           'cdk --version',
                           'cdk diff',
                       ]
                    },
                    build: {
                        commands: [
                            'npx cdk deploy --verbose --all --require-approval never'
                        ]
                    }
                }
            })
        })

        project.role?.addManagedPolicy(
            iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess")
        )
    }
}