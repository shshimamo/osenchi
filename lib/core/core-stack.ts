import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as cloudtrail from 'aws-cdk-lib/aws-cloudtrail';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';

export interface IBucketProps {
    inputName: string;
    outputName: string;
    logName: string;
}

export interface ITopicProps {
    topicName: string;
    emails: string[];
}

export interface ICoreProps extends cdk.StackProps {
    bucket: IBucketProps;
    topic: ITopicProps;
    detectionFuncName: string;
    deletionFuncName: string;
}

export class CoreStack extends cdk.Stack {
    /** 入力バケット */
    readonly inputBucket: s3.Bucket;

    /** 出力バケット */
    readonly outputBucket: s3.Bucket;

    /** メールトピック */
    readonly emailTopic: sns.Topic;

    /** 感情検出Lambda関数 */
    readonly detectionFunc: lambda.Function;

    /** オブジェクト削除Lambda関数 */
    readonly deletionFunc: lambda.Function;

    /** ステートマシン */
    readonly stateMachine: sfn.StateMachine;

    /**
     * スタックのインスタンスを生成します。
     *
     * @param {Construct} scope
     * @param {string} id
     * @param {ICoreProps} [props]
     * @memberof CoreStack
     */
    constructor(scope: Construct, id: string, props?: ICoreProps) {
        super(scope, id, props);

        if (props) {
            // バケット
            [this.inputBucket, this.outputBucket] = this.newBuckets(props.bucket);

            // トピック
            this.emailTopic = this.newEmailTopic(props.topic);

            // Lambda関数
            this.detectionFunc = this.newDetectSentiment(props.detectionFuncName);
            this.deletionFunc = this.newDeleteObject(props.deletionFuncName);

            // ワークフロー
            this.stateMachine = this.newStateMachine();

            // イベントルール
            this.newEventRule();
        }
    }

    /**
     * バケットの生成
     *
     * @private
     * @param {IBucketProps} bucketProps バケット情報
     * @returns {[s3.Bucket, s3.Bucket]}
     * @memberof CoreStack
     */
    private newBuckets(bucketProps: IBucketProps): [s3.Bucket, s3.Bucket] {
        const inputBucket = new s3.Bucket(this, 'InputBucket', {
            bucketName: bucketProps.inputName,
        });

        const outputBucket = new s3.Bucket(this, 'OutputBucket', {
            bucketName: bucketProps.outputName,
        });

        const logBucket = new s3.Bucket(this, 'LogBucket', {
            bucketName: bucketProps.logName,
        });

        const trail = new cloudtrail.Trail(this, 'Trail', {
            bucket: logBucket,
            isMultiRegionTrail: false,
        });
        const s3EventSelector: cloudtrail.S3EventSelector = {
            bucket: inputBucket,
        };
        trail.addS3EventSelector([s3EventSelector], {
            readWriteType: cloudtrail.ReadWriteType.WRITE_ONLY,
        });

        return [inputBucket, outputBucket];
    }

    /**
     * メールトピックの生成
     *
     * @private
     * @param {ITopicProps} topicProps トピック情報
     * @returns {sns.Topic}
     * @memberof CoreStack
     */
    private newEmailTopic(topicProps: ITopicProps): sns.Topic {
        const topic = new sns.Topic(this, 'Topic', {
            topicName: topicProps.topicName,
        });
        for (const email of topicProps.emails) {
            this.validateEmail(email);
            topic.addSubscription(new subscriptions.EmailSubscription(email));
        }
        return topic;
    }

    /**
     * メールアドレスを検証
     *
     * @param {string} email メールアドレス
     */
    private validateEmail(email: string): void {
        const regexp = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

        if (!email.match(regexp)) {
            throw new Error('Email parameter is invalid.');
        }
    }

    /**
     * 感情検出Lambda関数の生成
     *
     * @private
     * @param {string} functionName 関数名
     * @returns {lambda.Function}
     * @memberof CoreStack
     */
    private newDetectSentiment(functionName: string): lambda.Function {
        const func = new lambda.Function(this, 'DetectionFunc', {
            functionName: functionName,
            code: lambda.Code.fromAsset('functions/detect-sentiment', {
                exclude: ['*.ts'],
            }),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_12_X,
            timeout: cdk.Duration.minutes(10),
            environment: {
                DEST_BUCKET: this.outputBucket.bucketName,
            },
        });
        this.inputBucket.grantRead(func);
        this.outputBucket.grantWrite(func);

        const policy = new iam.PolicyStatement({
            resources: ['*'],
            actions: ['comprehend:BatchDetectSentiment'],
        });
        func.addToRolePolicy(policy);

        return func;
    }

    /**
     * オブジェクト削除Lambda関数の生成
     *
     * @private
     * @param {string} functionName 関数名
     * @returns {lambda.Function}
     * @memberof CoreStack
     */
    private newDeleteObject(functionName: string): lambda.Function {
        const func = new lambda.Function(this, 'DeletionFunc', {
            functionName: functionName,
            code: lambda.Code.fromAsset('functions/delete-object', {
                exclude: ['*.ts'],
            }),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_12_X,
        });
        this.inputBucket.grantDelete(func);

        return func;
    }

    /**
     * ステートマシンの生成
     *
     * @private
     * @returns {sfn.StateMachine}
     * @memberof CoreStack
     */
    private newStateMachine(): sfn.StateMachine {
        const sentimentTask = new tasks.LambdaInvoke(this, 'DetectSentiment', {
          lambdaFunction: this.detectionFunc,
        });
        const deleteTask = new tasks.LambdaInvoke(this, 'DeleteObject', {
          lambdaFunction: this.deletionFunc,
        });
        const successTask = new tasks.SnsPublish(this, 'SendSuccessMail', {
            topic: this.emailTopic,
            message: sfn.TaskInput.fromJsonPathAt('$.*'),
            subject: 'Osenchi Success',
        });
        const errorTask = new tasks.SnsPublish(this, 'SendErrorMail', {
            topic: this.emailTopic,
            message: sfn.TaskInput.fromJsonPathAt('$.*'),
            subject: 'Osenchi Error',
        });

        const fail = new sfn.Fail(this, 'Fail');
        errorTask.next(fail);

        const mainFlow = sentimentTask.next(deleteTask).next(successTask);

        const parallel = new sfn.Parallel(this, 'Parallel');
        parallel.branch(mainFlow);
        parallel.addCatch(errorTask, { resultPath: '$.error' });

        const stateMachine = new sfn.StateMachine(this, 'CoreStateMachine', {
            definition: parallel,
            timeout: cdk.Duration.minutes(30),
        });

        this.emailTopic.grantPublish(stateMachine.role);

        return stateMachine;
    }

    /**
     * イベントルールの生成
     *
     * @private
     * @returns {events.Rule}
     * @memberof CoreStack
     */
    private newEventRule(): events.Rule {
        const rule = new events.Rule(this, 'EventRule', {
            eventPattern: {
                source: ['aws.s3'],
                detailType: ['AWS API Call via CloudTrail'],
                detail: {
                    eventSource: ['s3.amazonaws.com'],
                    eventName: ['PutObject'],
                    requestParameters: {
                        bucketName: [this.inputBucket.bucketName],
                    },
                },
            },
        });

        const target = new targets.SfnStateMachine(this.stateMachine);
        rule.addTarget(target);

        return rule;
    }
}