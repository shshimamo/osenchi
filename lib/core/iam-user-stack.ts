import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';

export class IamUserStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        const accountId = cdk.Stack.of(this).account;
        const region = cdk.Stack.of(this).region;

        const codebuildRunPolicy = new iam.Policy(this, "codebuild-run", {
            policyName: "codebuild-run"
        })
        codebuildRunPolicy.addStatements(
            new iam.PolicyStatement({
                actions: ["codebuild:StartBuild", "codebuild:BatchGetBuilds"],
                effect: iam.Effect.ALLOW,
                resources: [`arn:aws:codebuild:${region}:${accountId}:project/*`]
            })
        )
        codebuildRunPolicy.addStatements(
            new iam.PolicyStatement({
                actions: ["logs:GetLogEvents"],
                effect: iam.Effect.ALLOW,
                resources: [
                    `arn:aws:logs:${region}:${accountId}:log-group:/aws/codebuild/*:*`
                ]
            })
        )
        const user = new iam.User(this, "githubactions", {
            userName: "githubactions",
        })
        user.attachInlinePolicy(codebuildRunPolicy)
    }
}