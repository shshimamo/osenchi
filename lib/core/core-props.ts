import { Node } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import { Context } from '../common/context';
import { ICoreProps } from './core-stack';

export class CoreProps {
    public static fromContext(node: Node): ICoreProps {
        const context = new Context(node);

        return {
            bucket: {
                inputName: context.service('input'),
                outputName: context.service('output'),
                logName: context.service('log'),
            },
            topic: {
                topicName: context.service('topic'),
                emails: context.get<string>('emails').split(','),
            },
            detectionFuncName: context.service('detect-sentiment'),
            deletionFuncName: context.service('delete-object'),
        };
    }
}