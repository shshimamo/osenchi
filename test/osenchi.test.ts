import * as cdk from 'aws-cdk-lib';
import { Template } from "aws-cdk-lib/assertions";
import { CoreProps } from '../lib/core/core-props';
import { CoreStack, ICoreProps } from '../lib/core/core-stack';
import { Context } from '../lib/common/context';

beforeAll(() => {
    Context.setEnvironment();
});

describe('Osenchi Snapshot tests', () => {
    test('Snapshot tests', () => {
        const app = new cdk.App();
        const props = CoreProps.fromContext(app.node);
        const stack = new CoreStack(app, 'Osenchi-Core', props);
        // スタックからテンプレート(JSON)を生成
        const template = Template.fromStack(stack).toJSON();
        expect(template).toMatchSnapshot();
    });
});