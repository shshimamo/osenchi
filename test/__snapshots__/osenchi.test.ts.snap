// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Osenchi Snapshot tests Snapshot tests 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CoreStateMachineB7B42254": Object {
      "DependsOn": Array [
        "CoreStateMachineRoleDefaultPolicyBB1EA2EF",
        "CoreStateMachineRole7A3F6625",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"Parallel\\",\\"States\\":{\\"Parallel\\":{\\"Type\\":\\"Parallel\\",\\"End\\":true,\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"ResultPath\\":\\"$.error\\",\\"Next\\":\\"SendErrorMail\\"}],\\"Branches\\":[{\\"StartAt\\":\\"DetectSentiment\\",\\"States\\":{\\"DetectSentiment\\":{\\"Next\\":\\"DeleteObject\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "DetectionFunc517A97A1",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"DeleteObject\\":{\\"Next\\":\\"SendSuccessMail\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "DeletionFunc033546BC",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"SendSuccessMail\\":{\\"End\\":true,\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish\\",\\"Parameters\\":{\\"TopicArn\\":\\"",
              Object {
                "Ref": "TopicBFC7AF6E",
              },
              "\\",\\"Message.$\\":\\"$.*\\",\\"Subject\\":\\"Osenchi Success\\"}}}}]},\\"SendErrorMail\\":{\\"Next\\":\\"Fail\\",\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish\\",\\"Parameters\\":{\\"TopicArn\\":\\"",
              Object {
                "Ref": "TopicBFC7AF6E",
              },
              "\\",\\"Message.$\\":\\"$.*\\",\\"Subject\\":\\"Osenchi Error\\"}},\\"Fail\\":{\\"Type\\":\\"Fail\\"}},\\"TimeoutSeconds\\":1800}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "CoreStateMachineRole7A3F6625",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "CoreStateMachineEventsRoleD78C356A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "CoreStateMachineEventsRoleDefaultPolicy5AC2504D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "CoreStateMachineB7B42254",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CoreStateMachineEventsRoleDefaultPolicy5AC2504D",
        "Roles": Array [
          Object {
            "Ref": "CoreStateMachineEventsRoleD78C356A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CoreStateMachineRole7A3F6625": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "CoreStateMachineRoleDefaultPolicyBB1EA2EF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TopicBFC7AF6E",
              },
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "DeletionFunc033546BC",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "DetectionFunc517A97A1",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "DeletionFunc033546BC",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "DetectionFunc517A97A1",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CoreStateMachineRoleDefaultPolicyBB1EA2EF",
        "Roles": Array [
          Object {
            "Ref": "CoreStateMachineRole7A3F6625",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeletionFunc033546BC": Object {
      "DependsOn": Array [
        "DeletionFuncServiceRoleDefaultPolicy9F90806E",
        "DeletionFuncServiceRoleBDC2D65F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "cd68757a640d547bd645f867e4cea8a5ef0d031c2dbf11bd19a7cc33387a037d.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "osenchi-shshimamo-dev-delete-object",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DeletionFuncServiceRoleBDC2D65F",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DeletionFuncServiceRoleBDC2D65F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DeletionFuncServiceRoleDefaultPolicy9F90806E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:DeleteObject*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "InputBucket3BF8630A",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeletionFuncServiceRoleDefaultPolicy9F90806E",
        "Roles": Array [
          Object {
            "Ref": "DeletionFuncServiceRoleBDC2D65F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DetectionFunc517A97A1": Object {
      "DependsOn": Array [
        "DetectionFuncServiceRoleDefaultPolicy0AD5E3F4",
        "DetectionFuncServiceRole05A10551",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a5a6f4129ea5bbbbd2584f4eda545bccd707441f6942fa31750b3f983d225469.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DEST_BUCKET": Object {
              "Ref": "OutputBucket7114EB27",
            },
          },
        },
        "FunctionName": "osenchi-shshimamo-dev-detect-sentiment",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DetectionFuncServiceRole05A10551",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 600,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DetectionFuncServiceRole05A10551": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DetectionFuncServiceRoleDefaultPolicy0AD5E3F4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:GetObject*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "InputBucket3BF8630A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "InputBucket3BF8630A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:Abort*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "OutputBucket7114EB27",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "OutputBucket7114EB27",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "comprehend:BatchDetectSentiment",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DetectionFuncServiceRoleDefaultPolicy0AD5E3F4",
        "Roles": Array [
          Object {
            "Ref": "DetectionFuncServiceRole05A10551",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventRule5A491D2C": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "eventName": Array [
              "PutObject",
            ],
            "eventSource": Array [
              "s3.amazonaws.com",
            ],
            "requestParameters": Object {
              "bucketName": Array [
                Object {
                  "Ref": "InputBucket3BF8630A",
                },
              ],
            },
          },
          "detail-type": Array [
            "AWS API Call via CloudTrail",
          ],
          "source": Array [
            "aws.s3",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "CoreStateMachineB7B42254",
            },
            "Id": "Target0",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "CoreStateMachineEventsRoleD78C356A",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "InputBucket3BF8630A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "osenchi-shshimamo-dev-input",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "LogBucketCC3B17E8": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "osenchi-shshimamo-dev-log",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "LogBucketPolicy900DBE48": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "LogBucketCC3B17E8",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LogBucketCC3B17E8",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "StringEquals": Object {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "LogBucketCC3B17E8",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "OutputBucket7114EB27": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "osenchi-shshimamo-dev-output",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TopicBFC7AF6E": Object {
      "Properties": Object {
        "TopicName": "osenchi-shshimamo-dev-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "Topicshshimamogmailcom5F773F6A": Object {
      "Properties": Object {
        "Endpoint": "shshimamo@gmail.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Trail022F0CF2": Object {
      "DependsOn": Array [
        "LogBucketPolicy900DBE48",
      ],
      "Properties": Object {
        "EnableLogFileValidation": true,
        "EventSelectors": Array [
          Object {
            "DataResources": Array [
              Object {
                "Type": "AWS::S3::Object",
                "Values": Array [
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "InputBucket3BF8630A",
                            "Arn",
                          ],
                        },
                        "/",
                      ],
                    ],
                  },
                ],
              },
            ],
            "ReadWriteType": "WriteOnly",
          },
        ],
        "IncludeGlobalServiceEvents": true,
        "IsLogging": true,
        "IsMultiRegionTrail": false,
        "S3BucketName": Object {
          "Ref": "LogBucketCC3B17E8",
        },
      },
      "Type": "AWS::CloudTrail::Trail",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
